runtime/compiler/p/codegen/PPCPrivateLinkage.cpp

```diff
@@ -1913,6 +1913,47 @@ int32_t TR::PPCPrivateLinkage::buildPrivateLinkageArgs(TR::Node
 	    addDependency(dependencies, NULL, (TR::RealRegister::RegNum)i, TR_VSX_SCALAR, cg());
       }
 
+
+
+   /*
+   PPC ABI --
+   AIX 32-bit BE:
+   AIX 64-bit BE:
+   Linux 64-bit BE: 
+   These follow aix-style linkage. 
+   There is a TOC-base register (gr2) that needs to be loaded from function descriptor of that function that is called. 
+   If the callee function is in libj9jitxx.so module, then the module's TOC-base can be found in the J9VMThread jitTOC field.
+   All CHelper functions are in libj9jitxx.so module.
+   The TOC can be loaded from the jitTOC field of J9VMThread (pointed to by metaDataRegister).
+  
+   Linux 32bit BE: Doesn't contain a TOC, the register gr2 is reserved.
+
+   Linux 64-bit LE: There exists a TOC (gr2), but there is no such function descriptor. 
+   The TOC-base set-up is defined by ABI to go through a function's global-entry (relocation). 
+   The global-entry address will be set up in gr12.
+
+   Registers r12, and r2 are always added to the dependency in the situations that they're used. 
+   R12 is added at all times, and is an unreserved register across all platforms. 
+   R2 is added to the dependency only if we're not dealing with a linux 32 bit platform. 
+   Ultimately, we don't need an assert or NULL check when searching for these registers within the dependency. 
+   */
+   if(linkage == TR_CHelper)
+   {
+      if(TR::Compiler->target.isLinux() && TR::Compiler->target.is64Bit() && TR::Compiler->target.cpu.isLittleEndian())
+      {
+         int32_t helperOffset = (callNode->getSymbolReference()->getReferenceNumber() - 1)*sizeof(intptrj_t);
+         generateTrg1MemInstruction(cg(), TR::InstOpCode::Op_load, callNode, dependencies->searchPreConditionRegister(TR::RealRegister::gr12), 
```

Comment:
Just curious: here we just set r12. There must be a code somewhere that does exactly the  same (loads helper address from the pseudo-TOC) and jumps to the helper. Which register is used there to contain the address? Should we just use r12 when we are jumping to the helper?

Reply:
That's done in the Trampoline. 

 openj9/runtime/compiler/runtime/Trampoline.cpp

#if defined(TR_TARGET_64BIT)
         // ld gr11, [grPTOC, 8*(cookie-1)]
         *(int32_t *)buffer = 0xe9700000 | (((cookie-1)*sizeof(intptrj_t)) & 0x0000ffff);
buffer += 4;

On Java JIT PPCLE64 the same instruction is done to load the target address into r11, in which we use to branch directly to the target. 

There is a version in OMR also that uses r12, which follows the ABI perfectly. 

I was told by @zl-wang that r11 is used for other reasons as well here.